source/AI.cpp

```diff
@@ -141,6 +143,45 @@ namespace {
 				bay.ship->SetCommands(Command::DEPLOY);
 	}
 
+	// Helper function for selecting the ships for formation commands.
+	vector<Ship *> GetShipsForFormationCommand(const PlayerInfo &player)
+	{
+		// Figure out what ships we are giving orders to
+		vector<Ship *> targetShips;
+		auto &selectedShips = player.SelectedShips();
+		bool fullFleet = selectedShips.empty();
+		if(fullFleet)
+		{
+			auto &playerShips = player.Ships();
+			targetShips.reserve(playerShips.size() - 1);
+			for(const shared_ptr<Ship> &it : player.Ships())
+				if(it.get() != player.Flagship() && !it->IsParked()
+					&& !it->IsDestroyed())
+					targetShips.push_back(it.get());
+		}
+		else
+		{
+			targetShips.reserve(selectedShips.size());
+			for(const weak_ptr<Ship> &it : selectedShips)
+			{
+				shared_ptr<Ship> ship = it.lock();
+				if(ship)
+					targetShips.push_back(ship.get());
+			}
+		}
+
+		// This can happen when the player only has a flagship or when
+		// all selected escorts are disabled.
+		if(targetShips.empty())
+		{
+			if(fullFleet)
+				Messages::Add("No ships in the fleet to change formation for.", Messages::Importance::High);
+			else
+				Messages::Add("No ships selected that can change formation.", Messages::Importance::High);
```

Comment:
warning or info might be more fitting?

Reply:
The player gives a command and then noting happens. This is quite an exceptional situation that shouldn't happen. I think info is too low, but warning sounds like a category that might also be fitting.

Reply:
I'll keep importance high, because the level below it is info, and I feel that that is too low.

Reply:
ok